### PR TITLE
Added View Count + Read TIme

### PR DESCRIPTION
**Pull request by ManishShah120**
## Description of the feature
- Added the _**View Count**_ feature which will display the number of times any blog-post has been viewed.
- Added _**Read Time**_ feature, which works perfectly, to calculate the approximate read time of any blog-post and display it on the Page.

## Additional Changes
- Added `utils.py` file where all the utility function will be stored for the project.
- Fixed the shape of the **Read More** button of the slider.
- Also added a Front-end UI feature for changing the previously uploaded image with a blog-post.

Fixes: #106 

## GIF [Feature]
view_read_count2

## ScreenShot
shot1

### CODE REVIEW

**Comment by ManishShah120** [comment]
Hey @Kajol-Kumari & @harsh-9in , please have a look I have tried my best to add the read time but couldn't do it, though the codes are present in this commit, If possible future contributors could solve the read_time issue, I have commented out the code which created problem while adding the read time feature. The View_count is working perfectly. If You see any changes which I can make which will result in working of the read time feature please suggest,I will be more than happy enough.

File: mysite/blog/views.py [hunk 1]
```diff
@@ -59,11 +62,26 @@ def Profileview(request,name):
...
+    def get_object(self):
+        obj = super().get_object()
+        obj.view_count += 1
+        obj.save()
+        return obj
+
+    # def get_context_data(self, *args, **kwargs):
+    #     context = super().get_context_data(**kwargs)
+    #     texts = self.object.content
+    #     print(texts)
```
Comment (Kajol-Kumari) [inline mysite/blog/views.py:74]: @ManishShah120 what outputs you are getting in here? I mean did you figure out which part of this function is actually not working as expected?

Comment (ManishShah120) [reply mysite/blog/views.py:74]:
```
    def get_context_data(self, *args, **kwargs):
        context = super().get_context_data(**kwargs)
        texts = self.object.content
        #print(texts)
        Read_Time=get_read_time(texts)
        print(type(Read_Time))
        print(Read_Time)
        #Post.objects.create(read_time=Read_Time)
        return context
```
By Uncommenting these lines this is the output I am getting:-
```
<class 'str'>
0:01:00
[04/Sep/2020 11:31:12] "GET /post/lorem-ipsum/ HTTP/1.1" 200 4665
```

Now I am trying to save `0:01:00` to the database, which is creating the error

---

When I am trying to save this value to the database:-
```
    def get_context_data(self, *args, **kwargs):
        context = super().get_context_data(**kwargs)
        texts = self.object.content
        #print(texts)
        Read_Time=get_read_time(texts)
        print(type(Read_Time))
        print(Read_Time)
        Post.objects.create(read_time=Read_Time)
        return context
```
This is error message I am getting:-
```
Environment:


Request Method: GET
Request URL: http://127.0.0.1:8000/post/lorem-ipsum/

Django Version: 3.0.8
Python Version: 3.8.2
Installed Applications:
['blog.apps.BlogConfig',
 'django.contrib.admin',
 'crispy_forms',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.staticfiles',
 'users.apps.UsersConfig']
Installed Middleware:
('whitenoise.middleware.WhiteNoiseMiddleware',
 'django.middleware.security.SecurityMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware')



Traceback (most recent call last):
  File "/home/manish/Videos/open-source/Blog-Env/lib/python3.8/site-packages/django/db/backends/utils.py", line 86, in _execute
    return self.cursor.execute(sql, params)
  File "/home/manish/Videos/open-source/Blog-Env/lib/python3.8/site-packages/django/db/backends/sqlite3/base.py", line 396, in execute
    return Database.Cursor.execute(self, query, params)

The above exception (NOT NULL constraint failed: blog_post.author_id) was the direct cause of the following exception:
  File "/home/manish/Videos/open-source/Blog-Env/lib/python3.8/site-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/home/manish/Videos/open-source/Blog-Env/lib/python3.8/site-packages/django/core/handlers/base.py", line 115, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/home/manish/Videos/open-source/Blog-Env/lib/python3.8/site-packages/django/core/handlers/base.py", line 113, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/manish/Videos/open-source/Blog-Env/lib/python3.8/site-packages/django/views/generic/base.py", line 71, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/manish/Videos/open-source/Blog-Env/lib/python3.8/site-packages/django/views/generic/base.py", line 97, in dispatch
    return handler(request, *args, **kwargs)
  File "/home/manish/Videos/open-source/Blog-Env/lib/python3.8/site-packages/django/views/generic/detail.py", line 107, in get
    context = self.get_context_data(object=self.object)
  File "/home/manish/Videos/open-source/Blogger/mysite/blog/views.py", line 78, in get_context_data
    Post.objects.create(read_time=Read_Time)
  File "/home/manish/Videos/open-source/Blog-Env/lib/python3.8/site-packages/django/db/models/manager.py", line 82, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/home/manish/Videos/open-source/Blog-Env/lib/python3.8/site-packages/django/db/models/query.py", line 433, in create
    obj.save(force_insert=True, using=self.db)
  File "/home/manish/Videos/open-source/Blogger/mysite/blog/models.py", line 22, in save
    super(Post, self).save(*args, **kwargs)
  File "/home/manish/Videos/open-source/Blog-Env/lib/python3.8/site-packages/django/db/models/base.py", line 745, in save
    self.save_base(using=using, force_insert=force_insert,
  File "/home/manish/Videos/open-source/Blog-Env/lib/python3.8/site-packages/django/db/models/base.py", line 782, in save_base
    updated = self._save_table(
  File "/home/manish/Videos/open-source/Blog-Env/lib/python3.8/site-packages/django/db/models/base.py", line 887, in _save_table
    results = self._do_insert(cls._base_manager, using, fields, returning_fields, raw)
  File "/home/manish/Videos/open-source/Blog-Env/lib/python3.8/site-packages/django/db/models/base.py", line 924, in _do_insert
    return manager._insert(
  File "/home/manish/Videos/open-source/Blog-Env/lib/python3.8/site-packages/django/db/models/manager.py", line 82, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/home/manish/Videos/open-source/Blog-Env/lib/python3.8/site-packages/django/db/models/query.py", line 1204, in _insert
    return query.get_compiler(using=using).execute_sql(returning_fields)
  File "/home/manish/Videos/open-source/Blog-Env/lib/python3.8/site-packages/django/db/models/sql/compiler.py", line 1392, in execute_sql
    cursor.execute(sql, params)
  File "/home/manish/Videos/open-source/Blog-Env/lib/python3.8/site-packages/django/db/backends/utils.py", line 100, in execute
    return super().execute(sql, params)
  File "/home/manish/Videos/open-source/Blog-Env/lib/python3.8/site-packages/django/db/backends/utils.py", line 68, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/home/manish/Videos/open-source/Blog-Env/lib/python3.8/site-packages/django/db/backends/utils.py", line 77, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/home/manish/Videos/open-source/Blog-Env/lib/python3.8/site-packages/django/db/backends/utils.py", line 86, in _execute
    return self.cursor.execute(sql, params)
  File "/home/manish/Videos/open-source/Blog-Env/lib/python3.8/site-packages/django/db/utils.py", line 90, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/home/manish/Videos/open-source/Blog-Env/lib/python3.8/site-packages/django/db/backends/utils.py", line 86, in _execute
    return self.cursor.execute(sql, params)
  File "/home/manish/Videos/open-source/Blog-Env/lib/python3.8/site-packages/django/db/backends/sqlite3/base.py", line 396, in execute
    return Database.Cursor.execute(self, query, params)

Exception Type: IntegrityError at /post/lorem-ipsum/
Exception Value: NOT NULL constraint failed: blog_post.author_id
```

Comment (Kajol-Kumari) [reply mysite/blog/views.py:74]: @ManishShah120 there is no need to store the expected view time in db. It's most of a frontend work. So, you just fetch the blog data from backend, calculate the expected read time and display it on frontend, no need to store it in db as we are not going to use it anywhere else. Can you please try this work flow, it should work then.

Comment (ManishShah120) [reply mysite/blog/views.py:74]: Sure.

**Comment by ManishShah120** [comment]
Hey @Kajol-Kumari , It Worked. A great thanx for suggesting me this idea.🤗 _If there are any conflicts do let me know._

**Comment by ManishShah120** [comment]
@harsh-9in, See if this can be merged.

**Comment by Kajol-Kumari** [comment]
@ManishShah120 can you please update the GIF now as I would like to see how the expected time is appearing.

**Comment by ManishShah120** [comment]
> @ManishShah120 can you please update the GIF now as I would like to see how the expected time is appearing on the UI.

Done @Kajol-Kumari

File: mysite/blog/migrations/0001_initial.py [hunk 2]
```diff
@@ -16,13 +16,20 @@ class Migration(migrations.Migration):
...
     operations = [
         migrations.CreateModel(
-            name='post',
+            name='Post',
             fields=[
                 ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                 ('title', models.CharField(max_length=100)),
                 ('content', models.TextField()),
                 ('date_posted', models.DateTimeField(default=django.utils.timezone.now)),
+                ('image', models.ImageField(blank=True, upload_to='blog_images')),
```
Comment (Kajol-Kumari) [inline mysite/blog/migrations/0001_initial.py:25]: @ManishShah120 we don't need this change for this PR. can you please remove it?

File: mysite/blog/migrations/0001_initial.py [hunk 3]
```diff
@@ -16,13 +16,20 @@ class Migration(migrations.Migration):
...
+                ('view_count', models.IntegerField(default=0)),
+                ('read_time', models.TextField()),
+                ('slug', models.SlugField(blank=True, null=True)),
```
Comment (Kajol-Kumari) [inline mysite/blog/migrations/0001_initial.py:28]: same for this, we don't need this change in this PR.

File: mysite/blog/migrations/0001_initial.py [hunk 4]
```diff
@@ -16,13 +16,20 @@ class Migration(migrations.Migration):
...
                 ('author', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to=settings.AUTH_USER_MODEL)),
             ],
+            options={
+                'ordering': ['-date_posted'],
```
Comment (Kajol-Kumari) [inline mysite/blog/migrations/0001_initial.py:32]: Same

**Review by Kajol-Kumari — CHANGES_REQUESTED** [review]
@ManishShah120 I can see few changes from other PRs(I have pointed out few of them) can you please update this PR to have the changes required for read+write feature only?

**Comment by ManishShah120** [comment]
> @ManishShah120 I can see few changes from other PRs(I have pointed out few of them) can you please update this PR to have the changes required for read+write feature only?

Done @Kajol-Kumari

File: mysite/blog/views.py [hunk 5]
```diff
@@ -59,11 +61,16 @@ def Profileview(request,name):
...
+        obj = super().get_object()
+        obj.view_count += 1
+        obj.save()
+        return obj
 
 
 class PostUpdateView(LoginRequiredMixin, UserPassesTestMixin, UpdateView):
     model = Post
-    fields = ['title', 'content']
+    fields = ['title', 'image', 'content']
```
Comment (Kajol-Kumari) [inline mysite/blog/views.py:73]: This is unnecessary change for this PR. can you please remove it?

Comment (ManishShah120) [reply mysite/blog/views.py:73]: This is required because what if any user wants to replace the previously uploaded image with the blog, If I remove this than any user from the front end will not be able to change the uploaded image associated with his blog-post. Though If this feature is not required I will remove it, but then users will only be able to upload the image once.
So, should I remove it. @Kajol-Kumari

Comment (ManishShah120) [reply mysite/blog/views.py:73]: I have updated the description of this PR, for this particular change.

Comment (Kajol-Kumari) [reply mysite/blog/views.py:73]: Thanks for letting me know.Let's keep these changes :)

**Review by Kajol-Kumari — APPROVED** [review]
Thanks for all the hard work @ManishShah120 . This can be merged now @harsh-9in

**Comment by ManishShah120** [comment]
> Thanks for all the hard work @ManishShah120 . This can be merged now @harsh-9in

Thank You @Kajol-Kumari
